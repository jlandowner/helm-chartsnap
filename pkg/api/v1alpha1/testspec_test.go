package v1alpha1

import (
	"testing"

	. "github.com/jlandowner/helm-chartsnap/pkg/snap/gomega"
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

func TestCharts(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Charts Suite")
}

var _ = Describe("TestSpec", func() {
	var _ = Describe("FromFile", func() {
		Context("when values.yaml has testSpec", func() {
			It("should load config", func() {
				var v SnapshotValues
				err := FromFile("testdata/testspec_values.yaml", &v)
				Expect(err).NotTo(HaveOccurred())
				Expect(v).To(MatchSnapShot())
			})
		})

		Context("when loading .chartsnap.yaml", func() {
			It("should load config", func() {
				var v SnapshotConfig
				err := FromFile("testdata/.chartsnap.yaml", &v)
				Expect(err).NotTo(HaveOccurred())
				Expect(v).To(MatchSnapShot())
			})
		})

		Context("when loading invalid yaml", func() {
			It("should not load config", func() {
				var v SnapshotValues
				err := FromFile("testdata/testspec_values_invalid.yaml", &v)
				Expect(err).Should(HaveOccurred())
				Expect(err.Error()).To(MatchSnapShot())
			})
		})

		Context("when loading not found", func() {
			It("should not load config", func() {
				var v SnapshotValues
				err := FromFile("testdata/notfound.yaml", &v)
				Expect(err).Should(HaveOccurred())
				Expect(err.Error()).To(MatchSnapShot())
			})
		})
	})

	var _ = Describe("Merge", func() {
		It("should merge dynamic fields", func() {
			cfg1 := SnapshotConfig{
				DynamicFields: []ManifestPath{
					{
						APIVersion: "v1",
						Kind:       "service",
						Name:       "chartsnap-app1",
						JSONPath: []string{
							"/spec/ports/0/targetPort",
						},
					},
					{
						APIVersion: "v1",
						Kind:       "Service",
						Name:       "chartsnap-app1",
						JSONPath: []string{
							"/spec/ports/1/targetPort",
						},
					},
				},
			}
			cfg2 := SnapshotConfig{
				DynamicFields: []ManifestPath{
					{
						APIVersion: "v1",
						Kind:       "service",
						Name:       "chartsnap-app1",
						JSONPath: []string{
							"/spec/ports/0/targetPort",
						},
					},
					{
						APIVersion: "v1",
						Kind:       "Pod",
						Name:       "chartsnap-app1-test-connection",
						JSONPath: []string{
							"/metadata/name",
						},
					},
				},
			}
			cfg1.Merge(cfg2)
			Expect(cfg1).To(MatchSnapShot())
		})

		It("should merge snapshot file extensions", func() {
			base := SnapshotConfig{
				SnapshotFileExt: "json",
			}
			overwrite := SnapshotConfig{
				SnapshotFileExt: "yaml",
			}
			base.Merge(overwrite)
			Expect(base.SnapshotFileExt).To(Equal("yaml"))
		})

		It("should override snapshot file extension if not set in current config", func() {
			base := SnapshotConfig{}
			overwrite := SnapshotConfig{
				SnapshotFileExt: "yaml",
			}
			base.Merge(overwrite)
			Expect(base.SnapshotFileExt).To(Equal("yaml"))
		})

		It("should merge snapshot version when target is empty", func() {
			base := SnapshotConfig{}
			overwrite := SnapshotConfig{
				SnapshotVersion: "v1",
			}
			base.Merge(overwrite)
			Expect(base.SnapshotVersion).To(Equal("v1"))
		})

		It("should override snapshot version when target has value", func() {
			base := SnapshotConfig{
				SnapshotVersion: "v2",
			}
			overwrite := SnapshotConfig{
				SnapshotVersion: "v1",
			}
			base.Merge(overwrite)
			Expect(base.SnapshotVersion).To(Equal("v1")) // overwrite の値で上書きされることを期待
		})
	})

})

func TestManifestPath_DynamicValue(t *testing.T) {
	type fields struct {
		Kind       string
		APIVersion string
		Name       string
		JSONPath   []string
		Base64     bool
	}
	tests := []struct {
		name   string
		fields fields
		want   string
	}{
		{
			name: "Test DynamicValue",
			fields: fields{
				Base64: false,
			},
			want: "###DYNAMIC_FIELD###",
		},
		{
			name: "Test DynamicValue Base64",
			fields: fields{
				Base64: true,
			},
			want: "IyMjRFlOQU1JQ19GSUVMRCMjIw==",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			v := &ManifestPath{
				Kind:       tt.fields.Kind,
				APIVersion: tt.fields.APIVersion,
				Name:       tt.fields.Name,
				JSONPath:   tt.fields.JSONPath,
				Base64:     tt.fields.Base64,
			}
			if got := v.DynamicValue(); got != tt.want {
				t.Errorf("ManifestPath.DynamicValue() = %v, want %v", got, tt.want)
			}
		})
	}
}
