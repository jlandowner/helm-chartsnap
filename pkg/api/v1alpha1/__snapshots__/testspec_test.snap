['TestSpec FromFile when loading .chartsnap.yaml should load config 1']
SnapShot = """
{
  \"DynamicFields\": [
    {
      \"Kind\": \"Secret\",
      \"APIVersion\": \"v1\",
      \"Name\": \"app1-cert\",
      \"JSONPath\": [
        \"/data/ca.crt\",
        \"/data/tls.crt\",
        \"/data/tls.key\"
      ],
      \"Base64\": true,
      \"Value\": \"\"
    }
  ],
  \"SnapshotFileExt\": \"\",
  \"SnapshotVersion\": \"\"
}
"""

['TestSpec FromFile when loading invalid yaml should not load config 1']
SnapShot = """
failed to decode file 'testdata/testspec_values_invalid.yaml': yaml: line 10: could not find expected ':'"""

['TestSpec FromFile when loading not found should not load config 1']
SnapShot = """
failed to open file 'testdata/notfound.yaml': open testdata/notfound.yaml: no such file or directory"""

['TestSpec FromFile when values.yaml has testSpec should load config 1']
SnapShot = """
{
  \"TestSpec\": {
    \"DynamicFields\": [
      {
        \"Kind\": \"Secret\",
        \"APIVersion\": \"v1\",
        \"Name\": \"app1-cert\",
        \"JSONPath\": [
          \"/data/ca.crt\",
          \"/data/tls.crt\",
          \"/data/tls.key\"
        ],
        \"Base64\": true,
        \"Value\": \"\"
      }
    ],
    \"SnapshotFileExt\": \"\",
    \"SnapshotVersion\": \"\"
  }
}
"""

['TestSpec Merge should merge dynamic fields 1']
SnapShot = """
{
  \"DynamicFields\": [
    {
      \"Kind\": \"service\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1\",
      \"JSONPath\": [
        \"/spec/ports/0/targetPort\"
      ],
      \"Base64\": false,
      \"Value\": \"\"
    },
    {
      \"Kind\": \"Pod\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1-test-connection\",
      \"JSONPath\": [
        \"/metadata/name\"
      ],
      \"Base64\": false,
      \"Value\": \"\"
    },
    {
      \"Kind\": \"service\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1\",
      \"JSONPath\": [
        \"/spec/ports/0/targetPort\"
      ],
      \"Base64\": false,
      \"Value\": \"\"
    },
    {
      \"Kind\": \"Service\",
      \"APIVersion\": \"v1\",
      \"Name\": \"chartsnap-app1\",
      \"JSONPath\": [
        \"/spec/ports/1/targetPort\"
      ],
      \"Base64\": false,
      \"Value\": \"\"
    }
  ],
  \"SnapshotFileExt\": \"\",
  \"SnapshotVersion\": \"\"
}
"""
